tests/core/io/test_stream_peer_tcp.h

```diff
@@ -0,0 +1,303 @@
+		REQUIRE(negative_ret != OK);
+
+		Error too_high_ret = spt->bind(70000, bind_ip);
+		REQUIRE(too_high_ret != OK);
+	}
+
+	SUBCASE("bind calls open and bind on NetSocket") {
+		int bind_port = 7890;
+		Error bind_ret = spt->bind(bind_port, bind_ip);
+		REQUIRE(bind_ret == OK);
+		REQUIRE(ns->is_open() == true);
+		REQUIRE(ns->host_ip == bind_ip);
+		REQUIRE(ns->host_port == bind_port);
+	}
+
+	SUBCASE("disconnect_from_host closes NetSocket") {
+		spt->disconnect_from_host();
+		REQUIRE(ns->is_open() == false);
+	}
+}
+
+TEST_CASE("[StreamPeerTCP] poll") {
+	Ref<MockNetSocket> ns = memnew(MockNetSocket);
+	Ref<StreamPeerTCP> spt = memnew(StreamPeerTCP);
+	IPAddress peer_ip = IPAddress("127.2.2.2");
+	int peer_port = 45878;
+	spt->accept_socket(ns, peer_ip, peer_port);
+	IPAddress bind_ip = IPAddress("127.0.0.1");
+	int bind_port = 9043;
+	Error bind_ret = spt->bind(bind_port, bind_ip);
+	REQUIRE(bind_ret == OK);
+	ns->_set_available_bytes(100);
+
+	SUBCASE("Unconnected status causes connect_to_host") {
+		Error poll_ret = spt->poll();
+		REQUIRE(poll_ret == OK);
+		REQUIRE(spt->get_status() == StreamPeerTCP::STATUS_CONNECTED);
+		REQUIRE(ns->is_open() == true);
+	}
+
+	SUBCASE("FIN causes disconnect_from_host") {
+		// This is the condition for FIN.
+		ns->_set_available_bytes(0);
+		Error fin_ret = spt->poll();
+		REQUIRE(fin_ret == OK);
+		REQUIRE(ns->is_open() == false);
+	}
+}
+
+TEST_CASE("[StreamPeerTCP] data") {
+	Ref<MockNetSocket> ns = memnew(MockNetSocket);
+	Ref<StreamPeerTCP> spt = memnew(StreamPeerTCP);
+	IPAddress peer_ip = IPAddress("127.5.4.3");
+	int peer_port = 8908;
+	spt->accept_socket(ns, peer_ip, peer_port);
+	IPAddress bind_ip = IPAddress("127.0.0.1");
+	int bind_port = 2039;
+	Error bind_ret = spt->bind(bind_port, bind_ip);
+	REQUIRE(bind_ret == OK);
+
+	SUBCASE("put_data an ascii-encoded stream") {
+		const char expected[] = "hello, world";
+		const size_t str_len = sizeof(expected);
+		uint8_t *in_data = new uint8_t[str_len];
+		memcpy(in_data, expected, str_len);
+		uint8_t *out_data = new uint8_t[str_len];
+		ns->_set_send_data(out_data);
+		uint8_t *observed_data = out_data;
+		Error put_ret = spt->put_data(in_data, str_len);
+		REQUIRE(put_ret == OK);
+		char observed[str_len];
+		convert_to_string(observed_data, observed, str_len);
+		REQUIRE(strcmp(observed, expected) == 0);
+		delete[] out_data;
+		delete[] in_data;
+	}
+
+	SUBCASE("get_data an ascii-encoded stream") {
+		const char expected[] = "I - too - say hello!";
+		const size_t str_len = sizeof(expected);
+		uint8_t *out_data = new uint8_t[str_len];
+		memcpy(out_data, expected, str_len);
+		ns->_set_recv_data(out_data);
+		uint8_t *in_data = new uint8_t[str_len];
+		uint8_t *observed_data = in_data;
+		Error get_ret = spt->get_data(in_data, str_len);
+		REQUIRE(get_ret == OK);
+		char observed[str_len];
+		convert_to_string(observed_data, observed, str_len);
+		REQUIRE(strcmp(observed, expected) == 0);
+		delete[] out_data;
+		delete[] in_data;
+	}
```

Comment:
Maybe you can check what I'm doing in https://github.com/godotengine/godot/pull/98043 PR.
I'm using to_ascii_buffer().ptr() on put_data calls and ptrw() on get_data calls.
Hope this helps.

```suggestion
	SUBCASE("put_data an ascii-encoded stream") {
		const String expected = "hello, world";
		Vector<uint8_t> send_data;
		send_data.resize_zeroed(expected.length());
		ns->_set_send_data(send_data.ptrw());

		REQUIRE_EQ(spt->put_data(expected.to_ascii_buffer().ptrw(), expected.length()), Error::OK);
		CHECK_EQ(send_data, expected.to_ascii_buffer());
	}

	SUBCASE("get_data an ascii-encoded stream") {
		const String expected = "I - too - say hello!";
		Vector<uint8_t> recv_data = expected.to_ascii_buffer();
		ns->_set_recv_data(recv_data.ptrw());

		Vector<uint8_t> in_data;
		in_data.resize(expected.length());
		REQUIRE_EQ(spt->get_data(in_data.ptrw(), expected.length()), Error::OK);
		CHECK_EQ(in_data, expected.to_ascii_buffer());
	}
```

Reply:
Thanks for this! I've actioned your suggestions 😄